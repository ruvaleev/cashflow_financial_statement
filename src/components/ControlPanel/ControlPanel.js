import React from  'react';

import ExpenseForm from '../ExpenseForm'
import IncomeForm from '../IncomeForm'
import UndoRedoPanel from '../UndoRedoPanel'

function ControlPanel({addExpense, addIncome, undo, redo}) {
  return (
    <div className='controlPanel'>
      <IncomeForm onSubmit={(e) => {
        e.preventDefault();
        addIncome({
          title: e.target.title.value,
          value: e.target.value.value,
          isPassive: e.target.isPassive.value == 'true'
        }),
        e.target.title.value = '',
        e.target.value.value = '';
      }}/>
      <ExpenseForm onSubmit={(e) => {
        e.preventDefault();
        addExpense({
          title: e.target.title.value,
          value: e.target.value.value
        }),
        e.target.title.value = '',
        e.target.value.value = '';
      }}/>

      <UndoRedoPanel redo={redo} undo={undo}/>
    </div>
  )
}

export default ControlPanel;
