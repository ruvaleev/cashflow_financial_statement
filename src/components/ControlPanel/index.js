import { connect } from 'react-redux';
import { ActionCreators } from 'redux-undo';

import { addExpense } from '../../redux/slices/expenses';
import { addIncome } from '../../redux/slices/incomes';
import ControlPanel from './ControlPanel';

const mapDispatchToProps = (dispatch) => ({
  addExpense: (data) => dispatch(addExpense(data)),
  addIncome: (data) => dispatch(addIncome(data)),
  undo: () => dispatch(ActionCreators.undo()),
  redo: () => dispatch(ActionCreators.redo())
})

export default connect(null, mapDispatchToProps)(ControlPanel);
