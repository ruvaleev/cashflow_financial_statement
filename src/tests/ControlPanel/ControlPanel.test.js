import React from 'react';
import { render, screen, fireEvent } from '@testing-library/react';
import '@testing-library/jest-dom/extend-expect';
import { Provider } from 'react-redux';
import configureStore from 'redux-mock-store';
import { ActionCreators } from 'redux-undo';

import Store from '../shared/Store';
import ControlPanel from '../../components/ControlPanel';

const mockStore = configureStore([]);
const leftClick = { button: 0 };

describe('ControlPanel', () => {
  let store;
  let component;
  
  beforeEach(() => {
    store = mockStore(Store());
    store.dispatch = jest.fn();

    component = render(
      <Provider store={store}>
        <ControlPanel/>
      </Provider>
    )
  });

  it('renders ControlPanel correctly', () => {
    expect(component).toMatchSnapshot();
  });

  it('dispatches Undo button click', () => {
    const undoButton = screen.queryByText('Отменить');
    fireEvent.click(undoButton, leftClick);
 
    expect(store.dispatch).toHaveBeenCalledTimes(1);
    expect(store.dispatch).toHaveBeenCalledWith(ActionCreators.undo());
  });

  it('dispatches Redo button click', () => {
    const redoButton = screen.queryByText('Повторить');
    fireEvent.click(redoButton, leftClick);
 
    expect(store.dispatch).toHaveBeenCalledTimes(1);
    expect(store.dispatch).toHaveBeenCalledWith(ActionCreators.redo());
  });
});
