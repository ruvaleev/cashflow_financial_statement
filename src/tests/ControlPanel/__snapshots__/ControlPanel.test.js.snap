// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`ControlPanel renders ControlPanel correctly 1`] = `
Object {
  "asFragment": [Function],
  "baseElement": <body>
    <div>
      <div
        class="controlPanel"
      >
        <div
          class="incomeForm"
        >
          <form
            class="bordered"
          >
            <div
              class="flex"
            >
              <label
                for="title"
              >
                Статья дохода
              </label>
              <input
                class="bordered withoutTopBorder"
                name="title"
                required=""
                type="text"
              />
            </div>
            <div
              class="flex"
            >
              <label
                for="value"
              >
                Сумма
              </label>
              <input
                class="bordered withoutTopBorder"
                name="value"
                type="number"
              />
            </div>
            <div>
              <label
                for="isPassive"
              >
                Доход пассивный?
              </label>
              <select
                class="bordered withoutTopBorder"
                name="isPassive"
              >
                <option
                  value="true"
                >
                  Да
                </option>
                <option
                  value="false"
                >
                  Нет
                </option>
              </select>
            </div>
            <button
              class="blackTitle blackButton font-18px"
            >
              Завести доход
            </button>
          </form>
        </div>
        <div
          class="expenseForm"
        >
          <form
            class="bordered"
          >
            <div
              class="flex"
            >
              <label
                for="title"
              >
                Статья расхода
              </label>
              <input
                class="bordered withoutTopBorder"
                name="title"
                required=""
                type="text"
              />
            </div>
            <div
              class="flex"
            >
              <label
                for="value"
              >
                Сумма
              </label>
              <input
                class="bordered withoutTopBorder"
                name="value"
                type="number"
              />
            </div>
            <button
              class="blackTitle blackButton font-18px"
            >
              Завести расход
            </button>
          </form>
        </div>
        <div
          class="undoRedoPanel"
        >
          <button
            class="blackTitle blackButton font-16px noLeftMargin"
          >
            Отменить
          </button>
          <button
            class="blackTitle blackButton font-16px noRightMargin"
          >
            Повторить
          </button>
        </div>
      </div>
    </div>
  </body>,
  "container": <div>
    <div
      class="controlPanel"
    >
      <div
        class="incomeForm"
      >
        <form
          class="bordered"
        >
          <div
            class="flex"
          >
            <label
              for="title"
            >
              Статья дохода
            </label>
            <input
              class="bordered withoutTopBorder"
              name="title"
              required=""
              type="text"
            />
          </div>
          <div
            class="flex"
          >
            <label
              for="value"
            >
              Сумма
            </label>
            <input
              class="bordered withoutTopBorder"
              name="value"
              type="number"
            />
          </div>
          <div>
            <label
              for="isPassive"
            >
              Доход пассивный?
            </label>
            <select
              class="bordered withoutTopBorder"
              name="isPassive"
            >
              <option
                value="true"
              >
                Да
              </option>
              <option
                value="false"
              >
                Нет
              </option>
            </select>
          </div>
          <button
            class="blackTitle blackButton font-18px"
          >
            Завести доход
          </button>
        </form>
      </div>
      <div
        class="expenseForm"
      >
        <form
          class="bordered"
        >
          <div
            class="flex"
          >
            <label
              for="title"
            >
              Статья расхода
            </label>
            <input
              class="bordered withoutTopBorder"
              name="title"
              required=""
              type="text"
            />
          </div>
          <div
            class="flex"
          >
            <label
              for="value"
            >
              Сумма
            </label>
            <input
              class="bordered withoutTopBorder"
              name="value"
              type="number"
            />
          </div>
          <button
            class="blackTitle blackButton font-18px"
          >
            Завести расход
          </button>
        </form>
      </div>
      <div
        class="undoRedoPanel"
      >
        <button
          class="blackTitle blackButton font-16px noLeftMargin"
        >
          Отменить
        </button>
        <button
          class="blackTitle blackButton font-16px noRightMargin"
        >
          Повторить
        </button>
      </div>
    </div>
  </div>,
  "debug": [Function],
  "findAllByAltText": [Function],
  "findAllByDisplayValue": [Function],
  "findAllByLabelText": [Function],
  "findAllByPlaceholderText": [Function],
  "findAllByRole": [Function],
  "findAllByTestId": [Function],
  "findAllByText": [Function],
  "findAllByTitle": [Function],
  "findByAltText": [Function],
  "findByDisplayValue": [Function],
  "findByLabelText": [Function],
  "findByPlaceholderText": [Function],
  "findByRole": [Function],
  "findByTestId": [Function],
  "findByText": [Function],
  "findByTitle": [Function],
  "getAllByAltText": [Function],
  "getAllByDisplayValue": [Function],
  "getAllByLabelText": [Function],
  "getAllByPlaceholderText": [Function],
  "getAllByRole": [Function],
  "getAllByTestId": [Function],
  "getAllByText": [Function],
  "getAllByTitle": [Function],
  "getByAltText": [Function],
  "getByDisplayValue": [Function],
  "getByLabelText": [Function],
  "getByPlaceholderText": [Function],
  "getByRole": [Function],
  "getByTestId": [Function],
  "getByText": [Function],
  "getByTitle": [Function],
  "queryAllByAltText": [Function],
  "queryAllByDisplayValue": [Function],
  "queryAllByLabelText": [Function],
  "queryAllByPlaceholderText": [Function],
  "queryAllByRole": [Function],
  "queryAllByTestId": [Function],
  "queryAllByText": [Function],
  "queryAllByTitle": [Function],
  "queryByAltText": [Function],
  "queryByDisplayValue": [Function],
  "queryByLabelText": [Function],
  "queryByPlaceholderText": [Function],
  "queryByRole": [Function],
  "queryByTestId": [Function],
  "queryByText": [Function],
  "queryByTitle": [Function],
  "rerender": [Function],
  "unmount": [Function],
}
`;
